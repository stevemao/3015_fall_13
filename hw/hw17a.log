This is pdfTeX, Version 3.141592-1.40.7 (MiKTeX 2.7) (preloaded format=pdflatex 2008.2.29)  9 NOV 2008 12:28
entering extended mode
**hw17a
(hw17a.tex
LaTeX2e <2005/12/01>
Babel <v3.8h> and hyphenation patterns for english, dumylang, nohyphenation, ge
rman, ngerman, french, loaded.
("C:\Program Files\MiKTeX 2.7\tex\latex\base\article.cls"
Document Class: article 2005/09/16 v1.4f Standard LaTeX document class
("C:\Program Files\MiKTeX 2.7\tex\latex\base\size11.clo"
File: size11.clo 2005/09/16 v1.4f Standard LaTeX file (size option)
)
\c@part=\count79
\c@section=\count80
\c@subsection=\count81
\c@subsubsection=\count82
\c@paragraph=\count83
\c@subparagraph=\count84
\c@figure=\count85
\c@table=\count86
\abovecaptionskip=\skip41
\belowcaptionskip=\skip42
\bibindent=\dimen102
)
\@input{header.aux}
(header.tex ("C:\Program Files\MiKTeX 2.7\tex\latex\amsfonts\amssymb.sty"
Package: amssymb 2002/01/22 v2.2d

("C:\Program Files\MiKTeX 2.7\tex\latex\amsfonts\amsfonts.sty"
Package: amsfonts 2001/10/25 v2.2f
\@emptytoks=\toks14
\symAMSa=\mathgroup4
\symAMSb=\mathgroup5
LaTeX Font Info:    Overwriting math alphabet `\mathfrak' in version `bold'
(Font)                  U/euf/m/n --> U/euf/b/n on input line 132.
)) (remark.sty)
(bussproofs.sty
Proof Tree (bussproofs) style macros. Version 0.6c.
\theLevel=\count87
\myMaxLevel=\count88
\myBoxA=\box26
\myBoxB=\box27
\myBoxC=\box28
\myBoxD=\box29
\myBoxLL=\box30
\myBoxRL=\box31
\thisAboveSkip=\dimen103
\thisBelowSkip=\dimen104
\newScoreStart=\dimen105
\newScoreEnd=\dimen106
\newCenter=\dimen107
\displace=\dimen108
\leftLowerAmt=\dimen109
\rightLowerAmt=\dimen110
\scoreHeight=\dimen111
\scoreDepth=\dimen112
)
\c@theorem=\count89
\c@conjecture=\count90
\c@corollary=\count91
\c@example=\count92
\c@fact=\count93
\c@lemma=\count94
\c@definition=\count95
\c@claim=\count96
\c@remark=\count97
\c@slogan=\count98
\c@axiom=\count99
\c@note=\count100
\c@problem=\count101
\c@exercise=\count102
)
No file hw17a.aux.
LaTeX Font Info:    Checking defaults for OML/cmm/m/it on input line 5.
LaTeX Font Info:    ... okay on input line 5.
LaTeX Font Info:    Checking defaults for T1/cmr/m/n on input line 5.
LaTeX Font Info:    ... okay on input line 5.
LaTeX Font Info:    Checking defaults for OT1/cmr/m/n on input line 5.
LaTeX Font Info:    ... okay on input line 5.
LaTeX Font Info:    Checking defaults for OMS/cmsy/m/n on input line 5.
LaTeX Font Info:    ... okay on input line 5.
LaTeX Font Info:    Checking defaults for OMX/cmex/m/n on input line 5.
LaTeX Font Info:    ... okay on input line 5.
LaTeX Font Info:    Checking defaults for U/cmr/m/n on input line 5.
LaTeX Font Info:    ... okay on input line 5.

Underfull \hbox (badness 10000) in paragraph at lines 6--7

 []

LaTeX Font Info:    Try loading font information for U+msa on input line 20.
("C:\Program Files\MiKTeX 2.7\tex\latex\amsfonts\umsa.fd"
File: umsa.fd 2002/01/19 v2.2g AMS font definitions
)
LaTeX Font Info:    Try loading font information for U+msb on input line 20.

("C:\Program Files\MiKTeX 2.7\tex\latex\amsfonts\umsb.fd"
File: umsb.fd 2002/01/19 v2.2g AMS font definitions
) [1



{C:/ProgramData/MiKTeX/2.7/pdftex/config/pdftex.map}] (hw17a.aux) ) 
Here is how much of TeX's memory you used:
 866 strings out of 95337
 8361 string characters out of 1184115
 52497 words of memory out of 1500000
 4099 multiletter control sequences out of 110000
 9119 words of font info for 35 fonts, out of 1200000 for 2000
 14 hyphenation exceptions out of 8191
 27i,4n,26p,238b,123s stack positions out of 5000i,500n,10000p,200000b,5000s
<C:/Program 
Files/MiKTeX 2.7/fonts/type1/bluesky/cm/cmbx10.pfb><C:/Program Files/MiKTeX 2.7
/fonts/type1/bluesky/cm/cmmi10.pfb><C:/Program Files/MiKTeX 2.7/fonts/type1/blu
esky/cm/cmr10.pfb><C:/Program Files/MiKTeX 2.7/fonts/type1/bluesky/cm/cmr8.pfb>
<C:/Program Files/MiKTeX 2.7/fonts/type1/bluesky/cm/cmsy10.pfb><C:/Program File
s/MiKTeX 2.7/fonts/type1/bluesky/cm/cmti10.pfb><C:/Program Files/MiKTeX 2.7/fon
ts/type1/bluesky/cm/cmtt10.pfb>
Output written on hw17a.pdf (1 page, 35964 bytes).
PDF statistics:
 34 PDF objects out of 1000 (max. 8388607)
 0 named destinations out of 1000 (max. 131072)
 1 words of extra memory for PDF output out of 10000 (max. 10000000)

