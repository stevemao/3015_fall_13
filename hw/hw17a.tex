\documentclass[11pt]{article}

\include{header}

\begin{document}
\homework{17}{11 November}


In class we started to present the unification algorithm.  I have
provided an implementation and some tests (in comments at the end.)
Your assignment is to extend the code by adding a product type
constructor to the datatype {\tt{Type}} and then to extend the rest of
the functions to handle this new type.  Note that this include all
functions that do pattern matching on the type constructors for
datatype {\tt{Type}}.

Recall:

\begin{definition}
Type terms $t_1$ and $t_2$ are {\em{unifiable}} if there is a substitution
$\sigma$ such that $\sigma t_1 = \sigma t_2$.
\end{definition}

Thus, a unifier is a substitution that makes the terms equal.  

The algorithm is specified as follows:

\begin{description}
\item{i.)} A variable (say $x$) unifies with any term (say $t$) as long as $x\not\in{} FV(t)$.
\item{ii.)} If terms $t_1$ and $t_2$ are not variables, they unify if
  they have the same constructor, and if their subterms unify.
\end{description}

\end{document}


% Local Variables:
% mode:latex
% comment-column:0
% comment-start: "% "
% compile-command: "pdflatex hw17a"
% fill-column:79
% End:




