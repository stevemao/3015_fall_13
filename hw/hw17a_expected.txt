*Type> :l Type
[1 of 1] Compiling Type             ( Type_Prod.hs, interpreted )
Ok, modules loaded: Type.
*Type>  test unify [(TyVar "a", TyVar "a")]
"[(a,a)] ---> []"
*Type>  test unify [(TyVar "a", TyVar "b")]
"[(a,b)] ---> [a := b]"
*Type>  test unify [(TyVar "b", TyVar "a")]
"[(b,a)] ---> [b := a]"
*Type>  test unify [(TyVar "a", Arrow (TyVar "a") (TyVar "b"))]
"[(a,(a -> b))] ---> *** Exception: unify: occurs check failure.
*Type>  test unify [(TyVar "a", Arrow (TyVar "b") (TyVar "b"))]
"[(a,(b -> b))] ---> [a := (b -> b)]"
*Type>  test unify [(Arrow (TyVar "b") (TyVar "b"), Arrow (TyVar "b") (TyVar "b"))]
"[((b -> b),(b -> b))] ---> []"
*Type>  test unify [(Arrow (TyVar "a") (TyVar "b"), Arrow (TyVar "b") (TyVar "a"))]
"[((a -> b),(b -> a))] ---> [a := b]"
*Type>  test unify [(Arrow (TyVar "b") (TyVar "a"), Arrow (TyVar "a") (TyVar "b"))]
"[((b -> a),(a -> b))] ---> [b := a]"
*Type>  test unify [(Arrow (TyVar "a") (Arrow (TyVar "b") (TyVar "c")), Arrow (Arrow (TyVar "c") (TyVar "b")) (TyVar "a"))]
"[((a -> (b -> c)),((c -> b) -> a))] ---> [b := c,a := (c -> b)]"
*Type>  test unify [(TyVar "a", Prod (TyVar "a") (TyVar "b"))]
"[(a,(a X b))] ---> *** Exception: unify: occurs check failure.
*Type>  test unify [(TyVar "a", Prod (TyVar "b") (TyVar "b"))]
"[(a,(b X b))] ---> [a := (b X b)]"
*Type>  test unify [(Prod (TyVar "b") (TyVar "b"), Prod (TyVar "b") (TyVar "b"))]
"[((b X b),(b X b))] ---> []"
*Type>  test unify [(Prod (TyVar "a") (TyVar "b"), Prod (TyVar "b") (TyVar "a"))]
"[((a X b),(b X a))] ---> [a := b]"
*Type>  test unify [(Prod (TyVar "b") (TyVar "a"), Prod (TyVar "a") (TyVar "b"))]
"[((b X a),(a X b))] ---> [b := a]"
*Type>  test unify [(Prod (TyVar "a") (Prod (TyVar "b") (TyVar "c")), Prod (Prod (TyVar "c") (TyVar "b")) (TyVar "a"))]
"[((a X (b X c)),((c X b) X a))] ---> [b := c,a := (c X b)]"
*Type>  test unify [(Arrow (TyVar "a") (Prod (TyVar "b") (TyVar "c")), Arrow (Prod (TyVar "c") (TyVar "b")) (TyVar "a"))]
"[((a -> (b X c)),((c X b) -> a))] ---> [b := c,a := (c X b)]"
*Type>  test unify [(Arrow (TyVar "b") (TyVar "b"), Prod (TyVar "b") (TyVar "b"))]
"[((b -> b),(b X b))] ---> *** Exception: unify: Arrow and Prod are not unifiable.
*Type>  test unify [(Prod (TyVar "b") (TyVar "b"), Arrow (TyVar "b") (TyVar "b"))]
"[((b X b),(b -> b))] ---> *** Exception: unify: Prod and Arrow are not unifiable.
*Type> 
